### PR TITLE
[MI-2304] Pipeline subscriptions

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. Updated API to support creating pipeline subscriptions.
2. Updated the UI for subscription modal to display pipeline event types.
3. Updated the UI of subscription card for for displaying pipeline subscriptions.

### CODE REVIEW

File: server/plugin/client.go [hunk 1]
```diff
@@ -134,6 +134,29 @@ func (c *client) callFormURLEncoded(url, path, method string, out interface{}, f
 	return c.Call(url, method, path, contentType, "", nil, out, formValues)
 }
 
+var publisherID = map[string]string{
+	constants.SubscriptionEventPullRequestCreated:                 "tfs",
```
Comment (ayusht2810) [inline server/plugin/client.go:138]: there is already a constant created for this. can't we use that here?

File: server/plugin/client.go [hunk 2]
```diff
@@ -134,6 +134,29 @@ func (c *client) callFormURLEncoded(url, path, method string, out interface{}, f
...
+	constants.SubscriptionEventPullRequestUpdated:                 "tfs",
+	constants.SubscriptionEventPullRequestCommented:               "tfs",
+	constants.SubscriptionEventPullRequestMerged:                  "tfs",
+	constants.SubscriptionEventCodePushed:                         "tfs",
+	constants.SubscriptionEventWorkItemCreated:                    "tfs",
+	constants.SubscriptionEventWorkItemUpdated:                    "tfs",
+	constants.SubscriptionEventWorkItemDeleted:                    "tfs",
+	constants.SubscriptionEventWorkItemCommented:                  "tfs",
+	constants.SubscriptionEventBuildCompleted:                     "tfs",
+	constants.SubscriptionEventReleaseAbandoned:                   "rm",
```
Comment (ayusht2810) [inline server/plugin/client.go:148]: same (we can create a constant if not present)

File: server/plugin/client.go [hunk 3]
```diff
@@ -160,10 +183,18 @@ func (c *client) CreateSubscription(body *serializers.CreateSubscriptionRequestP
 			Repository: body.Repository,
 			Branch:     body.TargetBranch,
 		}
+	case constants.ServiceTypePipelines:
+		payload.PublisherInputs = serializers.PublisherInputsPipelines{
+			ProjectID: project.ProjectID,
+		}
 	}
 
 	var subscription *serializers.SubscriptionValue
```
Comment (ayusht2810) [inline server/plugin/client.go:192]: declare this above where it is used

File: webapp/src/components/card/subscription/index.tsx [hunk 4]
```diff
@@ -17,77 +17,86 @@ type SubscriptionCardProps = {
...
+                        }
+                        labelExtraClassName='margin-left-5'
+                        value={pluginConstants.common.eventTypeMap[eventType as EventType] ?? ''}
+                    />
+                    <LabelValuePair
+                        labelIconClassName={`icon ${channelType === mm_constants.PRIVATE_CHANNEL ? 'icon-lock-outline' : 'icon-globe'} icon-label`}
+                        value={channelName}
+                    />
+                    <LabelValuePair
+                        labelIconClassName={'icon icon-account-outline icon-label'}
```
Comment (ayusht2810) [inline webapp/src/components/card/subscription/index.tsx:70]:
```suggestion
                        labelIconClassName='icon icon-account-outline icon-label'
```

Comment (ayusht2810) [reply webapp/src/components/card/subscription/index.tsx:70]: check everywhere

File: webapp/src/pluginConstants/common.ts [hunk 5]
```diff
@@ -36,10 +51,22 @@ export const serviceTypeMap: Record<EventType, string> = {
...
+    'ms.vss-release.release-abandoned-event': 'Release Abandoned',
+    'ms.vss-release.release-created-event': 'Release Created',
+    'ms.vss-release.deployment-approval-completed-event': 'Release deployment approval completed',
+    'ms.vss-release.deployment-approval-pending-event': 'Release deployment approval pending',
+    'ms.vss-release.deployment-completed-event': 'Release deployment completed',
+    'ms.vss-release.deployment-started-event': 'Release deployment started',
+    'ms.vss-pipelinechecks-events.approval-completed': 'Run stage approval completed',
+    'ms.vss-pipelines.stage-state-changed-event': 'Run stage state changed',
+    'ms.vss-pipelinechecks-events.approval-pending': 'Run stage waiting for approval',
+    'ms.vss-pipelines.run-state-changed-event': 'Run state changed',
```
Comment (ayusht2810) [inline webapp/src/pluginConstants/common.ts:64]: can't we reuse `eventTypePipelines` instead of again defining them?

**Review by Kshitij-Katiyar — APPROVED** [review]

File: webapp/src/pluginConstants/common.ts (hunk 5 above)
Comment (raghavaggarwal2308) [reply webapp/src/pluginConstants/common.ts:64]: it was extra removed it

File: webapp/src/components/card/subscription/index.tsx [hunk 6]
```diff
@@ -54,7 +54,7 @@ const SubscriptionCard = ({handleDeleteSubscrption, subscriptionDetails: {channe
                             <SVGWrapper
                                 width={12}
                                 height={12}
-                                viewBox='0 0 10 10'
+                                viewBox=' 0 0 10 10'
```
Comment (ayusht2810) [inline webapp/src/components/card/subscription/index.tsx:57]: why this change?

**Review by ayusht2810 — APPROVED** [review]

File: server/constants/oauth_config.go [hunk 7]
```diff
@@ -2,7 +2,7 @@ package constants
 
 const (
 	ResponseType        = "Assertion"
-	Scopes              = "vso.code_full vso.work_full"
+	Scopes              = "vso.build_execute vso.code_full vso.release_manage vso.work_full"
```
Comment (avas27JTG) [inline server/constants/oauth_config.go:5]: update this in READMEs also

detailed README can be updated later but info about scope is important

File: server/plugin/client.go [hunk 8]
```diff
@@ -135,6 +135,29 @@ func (c *client) callFormURLEncoded(url, path, method string, out interface{}, f
 	return c.Call(url, method, path, contentType, "", nil, out, formValues)
 }
 
+var publisherID = map[string]string{
```
Comment (avas27JTG) [inline server/plugin/client.go:138]: you should add some comments explaining why we needed this in brief so that in future other devs can get this

File: server/plugin/client.go [hunk 9]
```diff
@@ -161,8 +184,13 @@ func (c *client) CreateSubscription(body *serializers.CreateSubscriptionRequestP
 		},
 	}
 
+	baseURL := c.plugin.getConfiguration().AzureDevopsAPIBaseURL
+	if strings.Contains(body.EventType, "release") {
+		baseURL = strings.Replace(baseURL, "://", "://vsrm.", 1)
+	}
```
Comment (avas27JTG) [inline server/plugin/client.go:190]: create an util for this as it will be used frequently, and for now you can add a TODO comment on the util to use that at all such places where original baseURL is to be updated this way

File: server/serializers/subscriptions.go [hunk 10]
```diff
@@ -26,6 +26,10 @@ type PublisherInputsGeneric struct {
 	NotificationType             string `json:"notificationType,omitempty"`
 }
 
+type PublisherInputsPipelines struct {
+	ProjectID string `json:"projectId"`
+}
```
Comment (avas27JTG) [inline server/serializers/subscriptions.go:31]: I think this is not needed now as `generic` one is used for all

File: webapp/src/components/card/subscription/index.tsx [hunk 11]
```diff
@@ -17,83 +17,92 @@ type SubscriptionCardProps = {
...
-            <div className='project-details'>
-                <LabelValuePair
-                    label={
+const SubscriptionCard = ({handleDeleteSubscrption, subscriptionDetails: {channelType, eventType, serviceType, channelName, createdBy, targetBranch, repositoryName, pullRequestCreatedByName, pullRequestReviewersContainsName, pushedByName, mergeResultName, notificationTypeName, areaPath}, subscriptionDetails}: SubscriptionCardProps) => {
+    let serviceTypeIcon = pluginConstants.SVGIcons.boards;
+    if (serviceType === pluginConstants.common.repos) {
+        serviceTypeIcon = pluginConstants.SVGIcons.repos;
+    } else if (serviceType === pluginConstants.common.pipelines) {
+        serviceTypeIcon = pluginConstants.SVGIcons.pipelines;
+    }
```
Comment (avas27JTG) [inline webapp/src/components/card/subscription/index.tsx:26]: create a map like below
```
const serviceTypeIcon = {
  [pluginConstants.common.boards]: {
    icon: pluginConstants.SVGIcons.boards,
    viewBox: "0 0 16 16"
  },
....
[pluginConstants.common.pipelines]: {
    icon: pluginConstants.SVGIcons.pipelines,
    viewBox: "0 0 17 17"
  },
}
```

File: webapp/src/components/card/subscription/index.tsx [hunk 12]
```diff
@@ -17,83 +17,92 @@ type SubscriptionCardProps = {
...
+            <>
+                <div className='d-flex justify-content-between align-items-center mb-2'>
+                    <div className='d-flex align-item-center'>
                         <SVGWrapper
-                            width={12}
-                            height={12}
-                            viewBox='0 0 10 10'
+                            width={20}
+                            height={20}
+                            viewBox={serviceType === pluginConstants.common.pipelines ? ' 0 0 17 17' : ' 0 0 16 16'}
```
Comment (avas27JTG) [inline webapp/src/components/card/subscription/index.tsx:36]:
```suggestion
                            viewBox={pluginConstants.common.serviceTypeIcon[serviceType].viewBox}
```

File: webapp/src/components/card/subscription/index.tsx [hunk 13]
```diff
@@ -17,83 +17,92 @@ type SubscriptionCardProps = {
...
                         >
-                            {pluginConstants.SVGIcons.workEvent}
+                            {serviceTypeIcon}
```
Comment (avas27JTG) [inline webapp/src/components/card/subscription/index.tsx:38]:
```suggestion
                            {pluginConstants.common.serviceTypeIcon[serviceType].icon}
```

File: webapp/src/components/card/subscription/index.tsx [hunk 14]
```diff
@@ -17,83 +17,92 @@ type SubscriptionCardProps = {
...
+                    <LabelValuePair
+                        labelIconClassName={`icon ${channelType === mm_constants.PRIVATE_CHANNEL ? 'icon-lock-outline' : 'icon-globe'} icon-label`}
+                        value={channelName}
+                    />
+                    <LabelValuePair
+                        labelIconClassName={'icon icon-account-outline icon-label'}
+                        value={`Subscription created by ${createdBy}`}
+                    />
+                    {
+                        (areaPath || repositoryName || targetBranch || pullRequestCreatedByName || pullRequestReviewersContainsName || pushedByName || mergeResultName || notificationTypeName) && (
```
Comment (avas27JTG) [inline webapp/src/components/card/subscription/index.tsx:74]:
```suggestion
                        const showFilter = areaPath || repositoryName || targetBranch || pullRequestCreatedByName || pullRequestReviewersContainsName || pushedByName || mergeResultName || notificationTypeName
```

create this `const` above `return` and use that here as it has become long now

File: docs/oauth_setup.md [hunk 15]
```diff
@@ -9,7 +9,7 @@
 
     ![Screenshot from 2022-07-29 13-25-38](https://user-images.githubusercontent.com/100013900/181712472-d4faec27-a61c-4565-9e27-fc7156241b17.png)
 
-  - Under "Authorized scopes" check "Work items(full)" and "Code(full)".
+  - Under "Authorized scopes" check "Work items(full)", Build(read and execute), Release(read, write, execute and manage) and "Code(full)".
```
Comment (avas27JTG) [inline docs/oauth_setup.md:12]: update scope image as well

File: server/plugin/client.go [hunk 16]
```diff
@@ -135,6 +135,30 @@ func (c *client) callFormURLEncoded(url, path, method string, out interface{}, f
 	return c.Call(url, method, path, contentType, "", nil, out, formValues)
 }
 
+// publishedID is send in the payload while calling the Azure DevOps API and it varies according to the eventType
```
Comment (avas27JTG) [inline server/plugin/client.go:138]:
```suggestion
// publishedID is sent in the payload while calling the Azure DevOps API and it varies according to the eventType
```

File: server/plugin/client.go [hunk 17]
```diff
@@ -161,8 +185,10 @@ func (c *client) CreateSubscription(body *serializers.CreateSubscriptionRequestP
 		},
 	}
 
+	baseURL := c.plugin.updateBaseURLForReleaseEventTypes(c.plugin.getConfiguration().AzureDevopsAPIBaseURL, body.EventType)
+
```
Comment (avas27JTG) [inline server/plugin/client.go:189]:
```suggestion
```

File: docs/oauth_setup.md [hunk 18]
```diff
@@ -4,19 +4,21 @@
...
   - Under "Authorized scopes" check "Work items(full)", Build(read and execute), Release(read, write, execute and manage) and "Code(full)".
 
-    ![Screenshot from 2022-10-07 12-54-13](https://user-images.githubusercontent.com/55234496/194496403-dfd54566-ae6b-4daa-96c9-6a6c7e24c296.png)
+    ![image](https://user-images.githubusercontent.com/55234496/204722689-b3feef42-0aa5-42ff-b011-efe1a6dad4c6.png)
 
   - Click on "Create application".
   - On successful creation, you will be navigated to a page having App ID, App Secret, and Client Secret
 
-    ![Screenshot from 2022-10-07 13-02-54](https://user-images.githubusercontent.com/55234496/194498023-51eca666-7d58-47bf-80cf-74b4eecc1f04.png)
+    ![image](https://user-images.githubusercontent.com/55234496/204722316-dddac330-616a-4a2a-a144-eb4e2611265b.png)
```
Comment (avas27JTG) [inline docs/oauth_setup.md:21]: nice!

**Review by avas27JTG — APPROVED** [review]
